flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUpsertKeys.scala

```diff
@@ -186,6 +187,25 @@ class FlinkRelMdUpsertKeys private extends MetadataHandler[UpsertKeys] {
   }
 
   def getUpsertKeys(rel: Window, mq: RelMetadataQuery): JSet[ImmutableBitSet] = {
+    // If it's a ROW_NUMBER Rank, then the upsert keys are partition by key and row number.
+    if (rel.groups.length == 1) {
```

Comment:
Again, limiting the number of `groups` to 1 is weird. If in streaming, we can only support one `group` for now (I'm not 100 percent sure about it), you should at least add a comment about it for future maintainability.